pde_superresolution/WENO.py

```diff
@@ -0,0 +1,71 @@
+import numpy as np
+
+_p = (1 / 6) * np.array([
+	[2, -7, 11, 0, 0],
+	[0, -1, 5, 2, 0],
+	[0, 0, 2, 5, -1]
+])
+_B1 = np.array([
+	[1, -2, 1, 0, 0],
+	[0, 1, -2, 1, 0],
+	[0, 0, 1, -2, 1]
+])
+_B2 = np.array([
+	[1, -4, 3, 0, 0],
+	[0, 1, 0, -1, 0],
+	[0, 0, 3, -4, 1]
+])
+_d = np.array([[0.1, 0.6, 0.3]]).T
+
+class WENO(object):
+	"""
+	An implementation of 4th order finite volume WENO
+	"""
+
+	def __init__(self, dx, flux=lambda x: -0.5 * x ** 2, dflux=lambda x: -x):
+		"""Constructor.
+				Args:
+					dx: grid spacing
+					flux, dflux: callables that return the physical flux f(w) and its
+					derivative f'(w)
+		"""
+		self.flux = flux
+		self.dflux = dflux
+		self.eps = 1e-6
+		self.dx = dx
+
+	def flux_divergence(self, w: np.ndarray) -> np.ndarray:
+		"""
+		returns the WENO approximation of the divergence of the flux
+		"""
+		a = max(abs(self.dflux(w)))
+		f = self.flux(w)
+		v = 0.5 * (f + a * w)
+		u = np.roll(0.5 * (f - a * w), -1)
+
+		# right flux ######################
+		vs = np.stack([np.roll(v, k) for k in reversed(range(-2, 3))])
+
+		# polynomial reconstructions
+		ps = _p @ vs
```

Comment:
Inside google I will run this using Python 2 so I can use Beam -- so please use `np.matmul` instead for Python 2 compatibility

Reply:
that's easy